cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/batch/OpaqueDatasetCollection.java

```diff
@@ -108,12 +119,47 @@ public SparkCollection<T> join(JoinExpressionRequest joinRequest) {
   @Override
   public DataframeCollection toDataframeCollection(Schema schema) {
     StructType sparkSchema = DataFrames.toDataType(schema);
-    ExpressionEncoder<Row> encoder = RowEncoder.apply(sparkSchema);
+    ExpressionEncoder<Row> encoder = getRowEncoder(sparkSchema);
     Dataset<StructuredRecord> ds = (Dataset<StructuredRecord>) getDataset();
     MapFunction<StructuredRecord, Row> converter = r -> DataFrames.toRow(r, sparkSchema);
     return new DataframeCollection(schema, ds.map(converter, encoder),
         sec, jsc, sqlContext, datasetContext, sinkFactory, functionCacheFactory);
   }
 
+  /**
+   * This is required to handle breaking changes between spark 3.3.2 (Dataproc 2.1) to 3.5.1 (Dataproc 2.2).
+   * And we need to support both.
+   * Here we are trying to check if the new method introduced in 3.5.1 exists or not, and based on that we
+   * invoke the new method or the old one.
+   */
+  public static ExpressionEncoder<Row> getRowEncoder(StructType sparkSchema) {
+
+    try {
+      return ENCODER_CACHE.get(sparkSchema, () -> {
+        StringBuilder errorStrBuilder = new StringBuilder("Failed to load a suitable Encoder dynamically. Errors : ");
+        try {
+          Method encoderForMethod =  RowEncoder.class.getMethod("encoderFor", StructType.class);
+          Object agnosticEncoderObj = encoderForMethod.invoke(null, sparkSchema);
 
+          Method applyMethod = ExpressionEncoder.class.getMethod("apply",
+                         Class.forName("org.apache.spark.sql.catalyst.encoders.AgnosticEncoder"));
+          return (ExpressionEncoder<Row>)applyMethod.invoke(null, agnosticEncoderObj);
+        } catch (NoSuchMethodException | IllegalAccessException | InvocationTargetException |
+                 ClassNotFoundException e) {
+          errorStrBuilder.append(System.lineSeparator()).append(e.getMessage());
+        }
+
+        try {
+          // If it reaches here, meaning it should have the older method for 3.3.2 and lower.
+          Method applyMethod = RowEncoder.class.getMethod("apply", StructType.class);
```

Comment:
Do we also need to use `reflection` here?

This method is already present in spark 3.3.2, right?